src/main/java/spoon/pattern/internal/node/ElementNode.java

```diff
@@ -264,6 +265,9 @@ public void forEachParameterInfo(BiConsumer<ParameterInfo, RootNode> consumer) {
 		@SuppressWarnings("rawtypes")
 		CtElement clone = generator.getFactory().Core().create((Class) elementType.getMetamodelInterface().getActualClass());
 		generateSingleNodeAttributes(generator, clone, parameters);
+		if (this.templateElement instanceof CtArrayTypeReferenceImpl) {
+			((CtArrayTypeReferenceImpl<?>) clone).setDeclarationKind(((CtArrayTypeReferenceImpl<?>) this.templateElement).getDeclarationKind());
+		}
```

Comment:
The problem was here. The metadata of the declaration style was lost here.

Should we consider copying the metadata as well? We can make it generic and do it for all `ElementNode`s.

Reply:
```suggestion
		clone.setAllMetadata(this.templateElement.getAllMetadata());
```
Something like this can be done. What do you think?

Reply:
I'm not sure that's a good idea, propagating metadata in a clone is a major change to semantics.

@monperrus is there a conscious decision for not "cloning" metadata?

Reply:
> is there a conscious decision for not "cloning" metadata?

no, seems to me that it is reasonable to clone the metadata as well?

Reply:
> no, seems to me that it is reasonable to clone the metadata as well?

I'm unsure how it could affect e.g. gumtree-spoon-ast-diff, which uses metadata to map Gumtree nodes to Spoon nodes. And surely there is a reliance somewhere on not propagating metadata, so I'd label this as a breaking change if we use it.

Reply:
> I'm unsure how it could affect e.g. gumtree-spoon-ast-diff, which uses metadata to map Gumtree nodes to Spoon nodes.

I can check if gumtree-spoon is affected in any way.

> And surely there is a reliance somewhere on not propagating metadata

That could be an issue but our test cases are not able to catch that. I was also wondering that there are so many places where we clone elements. We should also discuss if they need to be changed as well. Otherwise, the definition of `clone` shall vary from place to place.

Reply:
> > I'm unsure how it could affect e.g. gumtree-spoon-ast-diff, which uses metadata to map Gumtree nodes to Spoon nodes.
> 
> I can check if gumtree-spoon is affected in any way.
> 
> > And surely there is a reliance somewhere on not propagating metadata
> 
> That could be an issue but our test cases are not able to catch that. I was also wondering that there are so many places where we clone elements.

I don't mean in Spoon, but in some other app or library that uses Spoon. It's a behavioral change, and thus potentially breaking to dependents as something may be relying on the old behavior.

> We should also discuss if they need to be changed as well. Otherwise, the definition of `clone` shall vary from place to place.

It might be good to add an explicit test to check that metadata is propagated in a clone.

> Nothing breaks on gumtree-spoon.

That's good.

> I added `topLevelClone.setAllMetadata(topLevelElement.getAllMetadata())` [here](https://github.com/INRIA/spoon/blob/master/src/main/java/spoon/support/visitor/equals/CloneHelper.java#L120) and all the tests still pass. Retaining metadata does not seem to affect anything.

In libraries we control. Which is good but not sufficient to not mark this a breaking change.

Reply:
> It might be good to add an explicit test to check that metadata is propagated in a clone.

```java
@Test
public void test_cloneCopiesMetadata() {
  String c1 = "public class A { }";
  CtType < ? > type = Launcher.parseClass(c1);
  type.putMetadata("meaning of life", 42);

  CtType < ? > typeClone = type.clone();

  assertEquals(1, typeClone.getAllMetadata().size());
}
```
This test case passed without the changes, which means that we have been cloning metadata as well, and this is [the line](https://github.com/INRIA/spoon/blob/673a5e93e8bf9af8894d5a99c76f1d61e1c23e49/src/main/java/spoon/support/visitor/clone/CloneBuilder.java#L60) that does it.

Cloning via `clone` API is different from cloning via `ElementNode`. Metadata was propagated for the former, however, it was ignored for the latter. I am unsure of the exact distinction between their semantics.

Reply:
@slarse I added the change for cloning element via template [here](https://github.com/INRIA/spoon/pull/4396).